pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+                }
+            });
+            let unlocked_val = self.total.saturating_sub(total);
+            self.total = total;
+            return unlocked_val;
+        }
+    }
+
+    #[derive(Clone, Default, Encode, Decode, RuntimeDebug)]
+    /// Snapshot of validator state at the start of the round for which they are selected
+    pub struct ValidatorSnapshot<AccountId, Balance> {
+        pub bond: Balance,
+        pub nominators: Vec<Bond<AccountId, Balance>>,
+        pub total: Balance,
+    }
+
+    impl<
+            A: Clone,
+            B: Copy + sp_std::ops::AddAssign + sp_std::ops::Add<Output = B> + sp_std::ops::SubAssign,
+        > From<Validator<A, B>> for ValidatorSnapshot<A, B>
+    {
+        fn from(other: Validator<A, B>) -> ValidatorSnapshot<A, B> {
+            ValidatorSnapshot {
+                bond: other.bond,
+                nominators: other.nominators.0,
+                total: other.bond + other.nomi_bond_total,
+            }
+        }
+    }
+
+    impl<AccountId: Ord, Balance: Ord> Eq for ValidatorSnapshot<AccountId, Balance> {}
+
+    impl<AccountId: Ord, Balance: Ord> PartialEq for ValidatorSnapshot<AccountId, Balance> {
+        fn eq(&self, other: &Self) -> bool {
+            self.nominators == other.nominators
+        }
+    }
+
+    impl<AccountId: Ord, Balance: Ord> Ord for ValidatorSnapshot<AccountId, Balance> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.total.cmp(&other.total)
+        }
+    }
+
+    impl<AccountId: Ord, Balance: Ord> PartialOrd for ValidatorSnapshot<AccountId, Balance> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+            // Some(self.total.cmp(&other.total))
+        }
+    }
+
+    #[derive(Encode, Decode, Clone, RuntimeDebug)]
+    pub struct Nominator<AccountId, Balance> {
+        pub nominations: OrderedSet<Bond<AccountId, Balance>>,
+        pub total: Balance,
+        pub active_bond: Balance,
+        pub unlocking: Vec<UnlockChunk<Balance>>,
+    }
+
+    impl<
+            AccountId: Ord + Clone,
+            Balance: Copy
+                + sp_std::ops::AddAssign
+                + sp_std::ops::Add<Output = Balance>
+                + sp_std::ops::SubAssign
+                + PartialOrd,
+        > Nominator<AccountId, Balance>
+    {
+        pub fn new(validator: AccountId, amount: Balance) -> Self {
+            Nominator {
+                nominations: OrderedSet::from(vec![Bond {
+                    owner: validator,
+                    amount,
+                }]),
+                total: amount,
+                active_bond: amount,
+                unlocking: Vec::new(),
+            }
+        }
+        pub fn add_nomination(&mut self, bond: Bond<AccountId, Balance>) -> bool {
+            let amt = bond.amount;
+            if self.nominations.insert(bond) {
+                self.total += amt;
+                self.active_bond += amt;
```

Comment:
use safe math or saturating operators :)

Reply:
:+1:  Moved to saturating_add()